### ISSUE TITLE
SimpleSamlPHP 1.19.1. Warning: session_create_id(): Failed to create new ID 

### ISSUE THREAD

**Issue by ineborako** [simplesamlphp/simplesamlphp#1478]
**Describe the bug**
We are using SimpleSamlPHP as a service provider for Magento 2.
PHP Version 7.4.18
SimpleSamlPHP version 1.19.1

This is an intermittent bug. Can't call authentication twice. The first launch is successful all others fails.

This is a code which is used inside Magento controller:

``` 
public function execute()
  {
        $callback_url = array_key_exists('callback_url', $_GET) ? $_GET['callback_url'] : '';
        $as = new \SimpleSAML\Auth\Simple('default-sp');
        $as->requireAuth([
            'ReturnTo' => '/test_sso/index/logincallback?callback_url='.$callback_url,
            'KeepPost' => FALSE,
            ''
        ]);
  }
```

**To Reproduce**
1. Cleanup browser cookie and call Magento controller - SimpleSamlPHP SSO is shown.
2. Click on the browser button 'Back' - the current page is Magento controller and it shows an error (see below).
3. All consequent attempts to make SSO call also fail until we clean up the browser cookie again
4. See error:

```
1 exception(s):
Exception #0 (Exception): Warning: session_create_id(): Failed to create new ID in /var/www/html/mg2/vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/SessionHandlerPHP.php on line 173
<pre>#1 session_create_id() called at [vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/SessionHandlerPHP.php:173]
#2 SimpleSAML\SessionHandlerPHP->newSessionId() called at [vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/Session.php:171]
#3 SimpleSAML\Session->__construct() called at [vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/Session.php:299]
#4 SimpleSAML\Session::getSessionFromRequest() called at [vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/Auth/Simple.php:53]
#5 SimpleSAML\Auth\Simple->__construct() called at [app/code/Tera/Sso/Controller/Index/Index.php:37]

```

**Expected behavior**
The SimpleSamlPHP SSO login form should be opened each time when we require authentication.

**Additional context**
The possible reason is the Magento session, because the standalone application works well (I created a simple page withe same code). I see in the code that SimpleSamlPHP tries to suspend the current session and replace it with its own. But it fails on an attempt to call standard function `session_create_id()`.

**Comment by ineborako** [comment]
I investigated a moment when `session_create_id()` when the function stops working. See screenshot below:
![изображение](https://user-images.githubusercontent.com/5718554/124984056-6eb84c00-e041-11eb-9c68-8b620fe4d33e.png)

**Comment by tvdijen** [comment]
Try and follow the docs;  https://simplesamlphp.org/docs/stable/simplesamlphp-sp#section_6

**Comment by ineborako** [comment]
Yes, I already tried - it didn't help. 
I thought that the active Magento session is the reason and I tried to suspend it:
```
session_write_close(); //Close Magento session
$as = new \SimpleSAML\Auth\Simple('default-sp');
```
But the result is the same - the first time it works, the others it doesn't.
I think that the real reason is in SessionHelperPHP.php and in Session.php. (See screenshot above).
Anyway, it is a bug - no matter how many times I execute this code it should give the same results.

**Comment by tvdijen** [comment]
Well, yes, but the `entire session_create_id()` is one big bug too, because it's not guaranteed for it to return an ID.. And it's behaviour is undocumented..  Real source for headache for a too long time now...
If you need a quick workaround, try and configure a different session store for SimpleSAMLphp, like memcached.

One more question though, are you integrating Magento and SimpleSAMLphp yourself? Or are you relying on some kind of Magento plugin?  We see these kind of issues a lot with plugins for Drupal/Wordpress/whatnot and usually the plugins are poorly written.

**Comment by ineborako** [comment]
Yes, we integrate SimpleSamlPHP with Magento ourselves.
I have several ideas:
1) using of try-catch for `session_create_id`. On exception - log a warning. Then fall back to `bin2hex(openssl_random_pseudo_bytes(16));`
![изображение](https://user-images.githubusercontent.com/5718554/125088910-8d1c5700-e0d6-11eb-80a0-f17021ce5fe4.png)

2) using configuration parameter that sets algorithm for session_id

**Comment by tvdijen** [comment]
That's already what we do in 1.19.x;
https://github.com/simplesamlphp/simplesamlphp/blob/simplesamlphp-1.19/lib/SimpleSAML/SessionHandlerPHP.php#L156-L183

session_create_id() will not throw an exception, but it will return false for unknown reasons.. I think it really depends on your PHP session configuration, but I'm not entirely sure how..

**Comment by tvdijen** [comment]
I think I'm starting to understand this a bit more.. `session_create_id()` will fail if a session already exists... Which makes sense..
The `Session::newSessionId()` shouldn't be called at all in this case..

**Comment by tvdijen** [comment]
@ineborako Would you mind trying this patch?  7583fefbc201ef438d68cd20ee2396b52f49cf96  It would be very helpful, thank you

**Comment by tvdijen** [comment]
Was closed in #1480
